exp/runtime/hooks/api/v1alpha1/discovery_types.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
+)
+
+// DiscoveryRequest represents the object of a discovery request.
```

Comment:
Note for a follow-up iteration. We should consider if/how to document hooks (e.g describing when the hook gets triggered etc etc.) so it surfaces in the Open API spec either commenting on the request type or on the func signature down in the file

Reply:
I'll open an issue / add a comment on an existing issue about this one and the next one